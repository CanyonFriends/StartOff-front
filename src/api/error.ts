export interface ErrorType {
  error_msg: string;
}

export const isFailed = <T>(arg: T | ErrorType): arg is ErrorType => {
  return (arg as ErrorType).error_msg !== undefined;
};
