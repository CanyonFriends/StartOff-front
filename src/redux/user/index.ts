import { createSlice } from '@reduxjs/toolkit';
import * as reducers from './reducer';
import { UserState } from './types';

export const initialState: UserState = {
  userId: '',
  nickname: '',
  email: '',
  error: '',
  isSignin: false,
};

export const userSlice = createSlice({
  initialState,
  name: 'user',
  reducers,
});

export const userReducer = userSlice.reducer;
export const { actions } = userSlice;
