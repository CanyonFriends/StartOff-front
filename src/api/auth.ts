import axios from 'axios';
import { SignupInfoType } from '../validator/signupValidator';
import { LoginInfoType } from '../validator/loginValidator';
import { ErrorType } from './error';

export interface SigninResponseType {
  access_token: string;
  uuid: string;
  email: string;
  nickname: string;
  user_id: string;
}

export interface SignupResponseType {
  user_id: string;
}

export interface LogoutRequestType {
  email: string;
  accessToken: string;
  uuid: string;
}

export interface LogoutResponseType {
  msg: string;
  success: boolean;
}

export const signupAPI = async ({ id, pw, nickname }: SignupInfoType): Promise<ErrorType | SignupResponseType> => {
  try {
    const response = await axios({
      method: 'POST',
      url: '/v1/signup',
      data: {
        nickname,
        email: id,
        password: pw,
      },
    });
    return response.data as SignupResponseType;
  } catch (error) {
    return { errorMsg: error.response.data.errorMsg } as ErrorType;
  }
};

export const signinAPI = async ({ id, pw }: LoginInfoType): Promise<ErrorType | SigninResponseType> => {
  try {
    const response = await axios({
      method: 'POST',
      url: '/v1/login',
      data: {
        email: id,
        password: pw,
      },
    });
    return response.data as SigninResponseType;
  } catch (error) {
    return { errorMsg: error.response.data.errorMsg } as ErrorType;
  }
};

export const logoutAPI = async ({
  email,
  accessToken,
  uuid,
}: LogoutRequestType): Promise<ErrorType | LogoutResponseType> => {
  try {
    const response = await axios({
      method: 'POST',
      url: '/v1/logout',
      headers: {
        Authorization: `Bearer ${accessToken}`,
      },
      data: {
        email,
        accessToken,
        uuid,
      },
    });
    return response.data as LogoutResponseType;
  } catch (error) {
    return { errorMsg: error.response.data.errorMsg } as ErrorType;
  }
};
