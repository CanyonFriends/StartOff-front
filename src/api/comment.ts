import { AxiosError } from 'axios';
import { ErrorType } from './error';
import axios from '../utils/axios';
import { CommentServerType } from '../@types/server';
import { commentServerType2ClientType } from '../converter/comment';

export interface CreateCommentRequest {
  postId: string;
  userId: string;
  content: string;
}

export interface UpdateCommentRequest {
  postId: string;
  userId: string;
  content: string;
  commentId: string;
}

export interface DeleteCommentRequest {
  postId: string;
  commentId: string;
}

export const createCommentAPI = async ({ postId, userId, content }: CreateCommentRequest) => {
  try {
    const response = await axios({
      method: 'POST',
      url: `/v1/posts/${postId}/comments`,
      data: {
        parend_id: null,
        user_id: userId,
        content,
      },
    });

    return commentServerType2ClientType(response.data as CommentServerType);
  } catch (error) {
    return { error_msg: (error as AxiosError).response?.data?.error_msg } as ErrorType;
  }
};

export const updateCommentAPI = async ({ postId, userId, content, commentId }: UpdateCommentRequest) => {
  try {
    const response = await axios({
      method: 'PUT',
      url: `/v1/posts/${postId}/comments/${commentId}`,
      data: {
        parend_id: null,
        user_id: userId,
        content,
      },
    });

    return commentServerType2ClientType(response.data as CommentServerType);
  } catch (error) {
    return { error_msg: (error as AxiosError).response?.data?.error_msg } as ErrorType;
  }
};

export const deleteCommentAPI = async ({ postId, commentId }: DeleteCommentRequest) => {
  try {
    await axios({
      method: 'DELETE',
      url: `/v1/posts/${postId}/comments/${commentId}`,
    });

    return true;
  } catch (error) {
    return { error_msg: (error as AxiosError).response?.data?.error_msg } as ErrorType;
  }
};
