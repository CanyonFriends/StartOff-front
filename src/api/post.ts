import axios from '../utils/axios';
import { ErrorType } from './error';
import { BoardServerType } from '../@types/server';
import { boardServerType2ClientType, createPostClientType2ServerType } from '../converter/post';
import { CreatePostClientType } from '../@types/client';

export const getCategoriesAPI = async () => {
  try {
    const response = await axios({
      method: 'GET',
      url: '/v1/categories',
    });

    return response.data as string[];
  } catch (error) {
    return { error_msg: error.response.data.error_msg } as ErrorType;
  }
};

export const getPostsAPI = async (page: number, size: number, category: string) => {
  try {
    const response = await axios({
      method: 'GET',
      url: `/v1/posts?page=${page}&size=${size}&category=${category}`,
    });

    return boardServerType2ClientType(response.data as BoardServerType);
  } catch (error) {
    return { error_msg: error.response.data.error_msg } as ErrorType;
  }
};

export const createPostAPI = async (post: CreatePostClientType) => {
  try {
    await axios({
      method: 'POST',
      url: '/v1/posts',
      data: {
        ...createPostClientType2ServerType(post),
      },
    });
    return true;
  } catch (error) {
    return { error_msg: error.response.data.error_msg } as ErrorType;
  }
};
