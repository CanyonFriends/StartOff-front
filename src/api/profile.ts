import { ProjectServerResponseType, SkillServerType } from '../@types/server';
import axios from '../utils/axios';
import { ErrorType } from './error';
import { ProfileType, SkillType } from '../@types/client';
import { profileResponse2Type } from '../converter/profile';
import { skillServerType2ClientType } from '../converter/skill';

export interface GetProfileRequest {
  userId: string;
}

export interface GetProfileResponse {
  baekjoon_id: string;
  blog_url: string;
  github_url: string;
  introduce: string;
  nickname: string;
  projects: ProjectServerResponseType[];
  user_skills: SkillServerType[];
}

export interface UpdateProfileIntroduceRequest {
  userId: string;
  introduce: string;
  nickname: string;
}

export interface UpdateProfileGithubRequest {
  userId: string;
  githubUrl: string;
}

export interface UpdateProfileBlogRequest {
  userId: string;
  blogUrl: string;
}

export interface UpdateUserSkillRequest {
  userId: string;
  skillName: string;
}

export interface DeleteUserSkillRequest {
  userId: string;
  skillId: string;
}

export const getProfileAPI = async ({ userId }: GetProfileRequest): Promise<ProfileType | ErrorType> => {
  try {
    const response = await axios({
      method: 'GET',
      url: `/v1/users/${userId}/profile`,
    });
    return profileResponse2Type(response.data as GetProfileResponse);
  } catch (error) {
    return { error_msg: error.response.data.error_msg } as ErrorType;
  }
};

export const updateProfileIntroduce = async ({
  userId,
  introduce,
  nickname,
}: UpdateProfileIntroduceRequest): Promise<boolean | ErrorType> => {
  try {
    await axios({
      method: 'PUT',
      url: `/v1/users/${userId}/introduce`,
      data: {
        introduce,
        nickname,
      },
    });
    return true;
  } catch (error) {
    return { error_msg: error.response.data.error_msg } as ErrorType;
  }
};

// TODO: 이름 변경
export const updateGithubIntroduce = async ({
  userId,
  githubUrl,
}: UpdateProfileGithubRequest): Promise<boolean | ErrorType> => {
  try {
    await axios({
      method: 'PUT',
      url: `/v1/users/${userId}/github-url`,
      data: {
        github_url: githubUrl,
      },
    });
    return true;
  } catch (error) {
    return { error_msg: error.response.data.error_msg } as ErrorType;
  }
};

export const updateBlogIntroduce = async ({
  userId,
  blogUrl,
}: UpdateProfileBlogRequest): Promise<boolean | ErrorType> => {
  try {
    await axios({
      method: 'PUT',
      url: `/v1/users/${userId}/blog-url`,
      data: {
        blog_url: blogUrl,
      },
    });
    return true;
  } catch (error) {
    return { error_msg: error.response.data.error_msg } as ErrorType;
  }
};

export const updateUserSkillAPI = async ({
  userId,
  skillName,
}: UpdateUserSkillRequest): Promise<SkillType | ErrorType> => {
  try {
    const response = await axios({
      method: 'PUT',
      url: `/v1/users/${userId}/skills`,
      data: {
        skill_name: skillName,
      },
    });
    return skillServerType2ClientType(response.data as SkillServerType);
  } catch (error) {
    return { error_msg: error.response.data.error_msg } as ErrorType;
  }
};

export const deleteUserSkillAPI = async ({ userId, skillId }: DeleteUserSkillRequest): Promise<boolean | ErrorType> => {
  try {
    await axios({
      method: 'DELETE',
      url: `/v1/users/${userId}/skills/${skillId}`,
    });
    return true;
  } catch (error) {
    return { error_msg: error.response.data.error_msg } as ErrorType;
  }
};
