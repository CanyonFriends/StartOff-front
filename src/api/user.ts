import axios from 'axios';
import { SignupInfoType } from '../validator/signupValidator';
import { LoginInfoType } from '../validator/loginValidator';
import { ErrorType } from './error';

export interface SigninResponseType {
  access_token: string;
  refresh_token: string;
  email: string;
  token_type: string;
  user_id: string;
}

export const signupAPI = async ({ id, pw, nickname }: SignupInfoType) => {
  try {
    const response = await axios({
      method: 'POST',
      url: '/v1/signup',
      data: {
        nickname,
        email: id,
        password: pw,
      },
    });
    return response;
  } catch (error) {
    return { error: error.response.data.errorMsg } as ErrorType;
  }
};

export const signinAPI = async ({ id, pw }: LoginInfoType): Promise<ErrorType | SigninResponseType> => {
  try {
    const response = await axios({
      method: 'POST',
      url: '/v1/login',
      data: {
        email: id,
        password: pw,
      },
    });
    return response.data as SigninResponseType;
  } catch (error) {
    return { error: error.response.data.errorMsg } as ErrorType;
  }
};
