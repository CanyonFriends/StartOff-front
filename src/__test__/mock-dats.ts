import { ErrorType } from '../api/error';
import { SigninResponseType, SignupResponseType } from '../api/user';

export const signinSuccessMockInfo: SigninResponseType = {
  access_token: 'access_token',
  refresh_token: 'refresh_token',
  email: 'email@email.com',
  token_type: 'bearer',
  user_id: '1',
};

export const signinFailMockInfo: ErrorType = {
  errorMsg: '에러 발생',
};

export const signupSuccessMockInfo: SignupResponseType = {
  user_id: '1',
};

export const signupFailMockInfo: ErrorType = {
  errorMsg: '에러 발생',
};
