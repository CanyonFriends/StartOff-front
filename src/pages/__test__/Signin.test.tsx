/**
 * @jest-environment jsdom
 */
import React from 'react';
import { fireEvent, waitFor } from '@testing-library/react';
import { render } from '../../test-utils';
import SigninPage from '../Signin';
import { signinFailMockInfo } from '../../__test__/mock-dats';
import { signinAPI } from '../../api/user';

jest.mock('../../api/user');

const signinMockAPI = signinAPI as jest.MockedFunction<typeof signinAPI>;

beforeEach(() => {
  jest.resetAllMocks();
});

describe('<Signin> 페이지', () => {
  it('snapshot 체크', () => {
    const component = render(<SigninPage />);
    expect(component.container).toMatchSnapshot();
  });

  it('input창들 초기값 비었는지 확인', () => {
    const component = render(<SigninPage />);
    const emailInput = component.getByLabelText('email') as HTMLInputElement;
    const passwordInput = component.getByLabelText('password') as HTMLInputElement;

    expect(emailInput.value).toBe('');
    expect(passwordInput.value).toBe('');
  });

  it('아이디 입력 안하면 경고', () => {
    const component = render(<SigninPage />);
    const loginButton = component.getByText('로그인');

    fireEvent.click(loginButton);
    component.getByText('아이디를 입력해주시기 바랍니다');
  });

  it('비밀번호 입력 안하면 경고', () => {
    const component = render(<SigninPage />);
    const emailInput = component.getByLabelText('email') as HTMLInputElement;
    const loginButton = component.getByText('로그인');

    fireEvent.change(emailInput, { target: { value: 'email@email.com' } });
    fireEvent.click(loginButton);
    component.getByText('비밀번호를 입력해주시기 바랍니다');
  });

  it('로그인 실패', () => {
    signinMockAPI.mockReturnValue(signinFailMockInfo);

    const component = render(<SigninPage />);
    const emailInput = component.getByLabelText('email') as HTMLInputElement;
    const passwordInput = component.getByLabelText('password') as HTMLInputElement;
    const loginButton = component.getByText('로그인');

    fireEvent.change(emailInput, { target: { value: 'email@email.com' } });
    fireEvent.change(passwordInput, { target: { value: 'email@email.com' } });
    fireEvent.click(loginButton);

    waitFor(() => component.getByText('에러 발생'));
  });
});
