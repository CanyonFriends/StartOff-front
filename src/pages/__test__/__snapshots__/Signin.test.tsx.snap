// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Signin> 페이지 snapshot 체크 1`] = `
<div>
  <div
    id="modal"
  />
  <main
    class="css-sapy0t"
  >
    <div
      class="css-2fxwiq"
    >
      <div
        aria-label=""
        class="css-17ymrk1"
      >
        <svg
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M12 0c-6.626 0-12 5.373-12 12 0 5.302 3.438 9.8 8.207 11.387.599.111.793-.261.793-.577v-2.234c-3.338.726-4.033-1.416-4.033-1.416-.546-1.387-1.333-1.756-1.333-1.756-1.089-.745.083-.729.083-.729 1.205.084 1.839 1.237 1.839 1.237 1.07 1.834 2.807 1.304 3.492.997.107-.775.418-1.305.762-1.604-2.665-.305-5.467-1.334-5.467-5.931 0-1.311.469-2.381 1.236-3.221-.124-.303-.535-1.524.117-3.176 0 0 1.008-.322 3.301 1.23.957-.266 1.983-.399 3.003-.404 1.02.005 2.047.138 3.006.404 2.291-1.552 3.297-1.23 3.297-1.23.653 1.653.242 2.874.118 3.176.77.84 1.235 1.911 1.235 3.221 0 4.609-2.807 5.624-5.479 5.921.43.372.823 1.102.823 2.222v3.293c0 .319.192.694.801.576 4.765-1.589 8.199-6.086 8.199-11.386 0-6.627-5.373-12-12-12z"
          />
        </svg>
      </div>
      <h1
        class="css-1hznjvn"
        color="#000000"
      >
        LOGIN
      </h1>
    </div>
    <form
      class="css-1ugbcgu"
    >
      <div
        class="css-1ydboqz"
      >
        <div
          class="css-1bechv"
        >
          <label
            class="css-1d3b76u"
            color="#404680"
            for="email"
          >
            email
          </label>
          <input
            aria-label="email"
            class="css-1556dao"
            id="email"
            name=""
            placeholder=""
            type="email"
            value=""
            width="100%"
          />
        </div>
      </div>
      <div
        class="css-1ydboqz"
      >
        <div
          class="css-1bechv"
        >
          <label
            class="css-1d3b76u"
            color="#404680"
            for="password"
          >
            password
          </label>
          <input
            aria-label="password"
            class="css-1556dao"
            id="password"
            name=""
            placeholder=""
            type="password"
            value=""
            width="100%"
          />
        </div>
      </div>
      <span
        class="css-rkz6lt"
      />
      <button
        class="css-7rx0q1"
        type="submit"
        width="100%"
      >
        로그인
      </button>
    </form>
    <div
      class="css-1mtiqxd"
    >
      <span
        class="css-1h4h9u4"
      >
        아직 계정이 없으시다면?
      </span>
      <div
        class="css-5jx5gk"
      >
        <a
          href="/signup"
        >
          계정 만들기
        </a>
      </div>
    </div>
    <button
      class="css-gp1fgf"
      type="submit"
      width="100%"
    >
      <div
        aria-label=""
        class="css-13sc1m2"
      >
        <svg
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M12 0c-6.626 0-12 5.373-12 12 0 5.302 3.438 9.8 8.207 11.387.599.111.793-.261.793-.577v-2.234c-3.338.726-4.033-1.416-4.033-1.416-.546-1.387-1.333-1.756-1.333-1.756-1.089-.745.083-.729.083-.729 1.205.084 1.839 1.237 1.839 1.237 1.07 1.834 2.807 1.304 3.492.997.107-.775.418-1.305.762-1.604-2.665-.305-5.467-1.334-5.467-5.931 0-1.311.469-2.381 1.236-3.221-.124-.303-.535-1.524.117-3.176 0 0 1.008-.322 3.301 1.23.957-.266 1.983-.399 3.003-.404 1.02.005 2.047.138 3.006.404 2.291-1.552 3.297-1.23 3.297-1.23.653 1.653.242 2.874.118 3.176.77.84 1.235 1.911 1.235 3.221 0 4.609-2.807 5.624-5.479 5.921.43.372.823 1.102.823 2.222v3.293c0 .319.192.694.801.576 4.765-1.589 8.199-6.086 8.199-11.386 0-6.627-5.373-12-12-12z"
          />
        </svg>
      </div>
      <span
        class="css-1xjc5z"
      >
        Github로 로그인
      </span>
    </button>
  </main>
</div>
`;
